Bluetooth: Controller: Fix conn timeout due to delayed conn upd ind

cvinayak opened the pull request:
Add connection parameter update to throughput test to cover
any connection timeout performing connection update while
subjected to high throughput scenarios.

Unlike legacy Zephyr Controller's double headed Tx queue
linked list, the new split Zephyr Controller's new LLCP does
not place a control PDU at the head of the Tx queue causing
control PDUs to be delayed on air.

Pause the data enqueue until pending PDUs are transmitted
across before enqueuing LE Connection Update Ind PDU, and
thereafter resume other PDUs being enqueued again.

There can still be connection timeout if enqueued PDUs are
not being transmitted/acked during repeated retransmissions.

Below steps will be used to ensure OTA DFU with high throughput and connection updates is handled gracefully, without connection timeout or assertions in used HCI Controller or Peripheral Device under test:
```
# Build and flash uart_hci onto nRF52840-DK
west build -p auto -d build_uart_hci/ --board nrf52840dk/nrf52840 samples/bluetooth/hci_uart
west flash -d build_uart_hci

# Build device firmware A
west build -p auto -d build_A --board nrf54l15dk/nrf54l15/cpuapp samples/subsys/mgmt/mcumgr/smp_svr/ -DCONFIG_DEBUG=y -DCONFIG_MCUMGR_TRANSPORT_BT_PERM_RW=y -DCONFIG_BT_DEVICE_NAME=\"firmwareA\" -DEXTRA_CONF_FILE="overlay-bt.conf" --sysbuild
# Build and flash device firmware B onto nRF54l15-DK
west build -p auto -d build_B --board nrf54l15dk/nrf54l15/cpuapp samples/subsys/mgmt/mcumgr/smp_svr/ -DCONFIG_DEBUG=y -DCONFIG_MCUMGR_TRANSPORT_BT_PERM_RW=y -DCONFIG_BT_DEVICE_NAME=\"firmwareB\" -DEXTRA_CONF_FILE="overlay-bt.conf" --sysbuild
west flash -d build_B

# Attach HCI
sudo btattach -S 1000000 -B /dev/ttyACM1 &
# Start FOTA to firmware A
sudo mcumgr --conntype ble --hci 1 --timeout 60 --connstring peer_name=\"firmwareA\" image upload build_A/smp_svr/zephyr/zephyr.bin
# Start FOTA to firmware B
sudo mcumgr --conntype ble --hci 1 --timeout 60 --connstring peer_name=\"firmwareB\" image upload build_B/smp_svr/zephyr/zephyr.bin
```

Script used for testing:
```
#! /bin/sh

cnt=1
while [ $cnt -le 1000 ]
do
	echo "Count $cnt"
	/home/jack/go/bin/mcumgr --conntype ble --hci 1 --timeout 60 --connstring peer_name="firmwareB" image upload build_A/smp_svr/zephyr/zephyr.signed.bin
	((cnt++))
done
```

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -17,6 +17,10 @@
 static uint8_t phy_param_idx;
 
+/* Count down number of metrics intervals before performing a param update */
+#define PARAM_UPDATE_COUNTDOWN PHY_UPDATE_COUNTDOWN
```
IT'S THE PHYNAL COUNTDOWN
![image](https://github.com/user-attachments/assets/ce952798-7baa-463b-806d-25ab1051a6e2)

(sorry, couldn't help myself)

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -178,6 +182,34 @@ static void write_cmd_cb(struct bt_conn *conn, void *user_data)
+		static uint8_t param_update_count;
+		int err;
+
+		if (param_update_countdown--) {
```
```suggestion
		if ((param_update_countdown--) != 0U) {
```

cvinayak replied:
Done

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -178,6 +182,34 @@ static void write_cmd_cb(struct bt_conn *conn, void *user_data)
+		param_update_countdown = PARAM_UPDATE_COUNTDOWN;
+
+		err = bt_conn_le_param_update(conn, &update_params[param_update_count & 0x1]);
+		if (err) {
```
```suggestion
		if (err != 0) {
```

cvinayak replied:
done

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -256,6 +288,9 @@ static void connected(struct bt_conn *conn, uint8_t conn_err)
 		phy_param_idx = 0U;
 	}
+
+	/* Offset param update by 1 metric collection after PHY update */
+	param_update_countdown = PARAM_UPDATE_COUNTDOWN + 1U;
```
I don't understand the comment, can you elaborate? What is a metric collection?

cvinayak replied:
https://github.com/zephyrproject-rtos/zephyr/pull/92572/files#diff-25e6d698734de351e15e26aa17ef532b57cbd3f591decaf457ab340eb16c44f0R141-R142

Every 1 second the accumulated data size is used for the throughput it calculation. PHY update is performed in reference to this calculation interval, and connection update is offset by 1 of this interval.

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -17,6 +17,10 @@
 static uint32_t phy_update_countdown;
 static uint8_t phy_param_idx;
 
+/* Count down number of metrics intervals before performing a param update */
+#define PARAM_UPDATE_COUNTDOWN PHY_UPDATE_COUNTDOWN
```
Also, this feels like even more test code that isn't related to GATT :( Would prefer that we test things like this in BSIM, instead of polluting a GATT sample with PHY tests (personal opinion)

cvinayak replied:
This may not be related to GATT but a typical user application requiring a GATT Write without Response will perform both PHY update and Connection Update to tune the throughput. In that sense, is a sample that demonstrate GATT Write without Response, PHY Update and Connection Update (there are no samples for these procedures currently in the upstream repo!).

cvinayak committed:
Bluetooth: Controller: Remove duplicate define CONN_DATA_BUFFERS

Remove duplicate define CONN_DATA_BUFFERS and use
CONFIG_BT_BUF_ACL_TX_COUNT instead.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Param update implementation review rework

Updated the test with review rework, and added comments.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Add param update to throughput test

Add connection parameter update to throughput test to cover
any connection timeout performing connection update while
subjected to high throughput scenarios.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: Fail throughput test with increased Tx count

Make the throughput test to fail when increasing Tx buffers
that will delay a LE Connection Ind PDU from being
transmitted on air due to previously queued Data PDUs.

Unlike legacy Zephyr Controller's double headed Tx queue
linked list, the new split Zephyr Controller's new LLCP does
not place a control PDU at the head of the queue causing
control PDUs to be delayed on air.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix conn timeout due to delayed conn upd ind

Unlike legacy Zephyr Controller's double headed Tx queue
linked list, the new split Zephyr Controller's new LLCP does
not place a control PDU at the head of the Tx queue causing
control PDUs to be delayed on air.

Pause the data enqueue until pending PDUs are transmitted
across before enqueuing LE Connection Update Ind PDU, and
thereafter resume other PDUs being enqueued again.

There can still be connection timeout if enqueued PDUs are
not being transmitted/acked during repeated retransmissions.

Updated bap_unicast_audio_acl_disconnect test to not use
random seed, to be similar to unicast_client test; to get
it to pass without overlapping state/roles casing connection
failed to be established after the changes in this PR.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix LLCP event_counter used during current event

Fix LLCP to use event_counter in the current event to
compare instant when trying to check instant passed on
reception of PDU with instant.

Fixes regression in commit e3e410aeed4e ("Bluetooth: Controller:
Fix LLCP event_counter value used during prepare").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix stuck forced continuation under throughput

Fix stuck forced continuation of connection event under high
throughput scenario. At near supervision timeout the force
flag is set, but in the next connection event if full
connection interval is used for data transmission and a
subsequent request to abort happens it was not honoured
causing the tx-rx chain to keep continuing until supervision
timeout.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (dismissed):
Looks OK to me, but with a few minor comments

thoh-ot reviewed (dismissed):
QACK.

I feel a better solution to pausing data would be to port the `CONFIG_BT_CTLR_PHY_UPDATE_IND_TX_ACKED_INSTANT_OVERRIDE` feature from PHY Update and then correct the instant in the LLL when transmitting the actual PDU.

cvinayak commented:
> correct the instant in the LLL 

I thought we were against modification of PDUs in the LLL once ULL has enqueued it.

is `CONFIG_BT_CTLR_PHY_UPDATE_IND_TX_ACKED_INSTANT_OVERRIDE` in the upstream repo?

thoh-ot commented:
> > correct the instant in the LLL
> 
> I thought we were against modification of PDUs in the LLL once ULL has enqueued it.
>
> is `CONFIG_BT_CTLR_PHY_UPDATE_IND_TX_ACKED_INSTANT_OVERRIDE` in the upstream repo?

Ahh, correct it is only in our downstream repo. It does also feel a bit hackish to modify the PDU in the LLL, but there we have LLL specific information, that can try to correct stale instant values. Dunno the reason why we never upstreamed that feature.

cvinayak committed:
Bluetooth: Controller: Fix missing radio tmr status reset

Fix missing radio tmr status reset on radio_disable call,
without this which disables PPI/DPPI caused under race
conditions the radio to transition into Tx or Rx state and
put the Controller's state out of sync with the Radio
peripherals internal state. The symptoms being that the
Central role was not transmitting in the next interval;
also, caused the defer iterations to reach maximum events
being enqueued in the pipeline, and manifesting as large
CPU overheads in redundant prepare of deferred events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing radio status reset

Fix missing radio status reset on early abort of central
and peripheral prepare.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Disable connection event continuation on overlap

Disable connection event continuation on overlap with same
connection's next event. This implementation has bugs and
needs to be fixed in subsequent commits.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Introduce prepare deferred feature

Introduce prepare being deferred due to previous events
desire to continue. In such case the deferred prepare param
will have the defer flag set.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix prepare pipeline for overlapping events

Fix prepare pipeline processing of overlapping events by
including ticks_at_expire value to uniquely identify
overlapping events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Added a regression fix in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/92572/commits/027e2752ede43866e3f0ffb29364ef3359d80d56

Earlier control procedure fix for event counter use in CIS create is required for Conn Update too. BSIM CI test now covers iteration of conn update which exposes the bug without the fix commit.

cvinayak committed:
Bluetooth: Controller: Fix window widening on connection update

Fix window widening on connection update to current window
widening periodic at the instant, and use new periodic
window widening for subsequent new connection interval.

Relates to commit 247037bd3e2a ("Bluetooth: Controller: Fix
incorrect elapsed events value").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix drift compensation on connection update

Fix drift compensation applied late under high throughput at
after the instant has passed for a connection update.

Workaround test failures due to Zephyr Controller scheduling
overlap causing Periodic Advertising Sync timeouts, ACL
connection failed to be established or Controller assertion
when having latency at the CIS setup instant.

Workaround by using `-start_offset` to ensure that the ACL
established does not overlap the sink being established.

Relates to commit 5f852998d046 ("tests: bsim: Bluetooth:
Workaround Zephyr Controller scheduling overlap").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix prepare pipeline overflow in observer role

Fix long standing bug in processing resume events that cause
prepare pipeline overflow due to incorrectly enqueueing up
preemptor events as duplicate events in the pipeline where
being aborted before currently active event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
Blocked by #93908. Needs rebase and `west update` there after.

cvinayak committed:
Bluetooth: Controller: Fix event overhead check for single timer use

Single timer use needs additional event overhead to start
radio to ensure the software switch implementation executing
in radio completion ISR can correctly set the tIFS switching
after a maximum radio ISR latency.

The latency value shall be the used additional overhead so
that there is no spurious tIFS switch before radio ISR
executes.

The additional event overhead value is not considered in the
event overhead check.

Relates to commit 37bf99eee743 ("Bluetooth: Controller: Fix
radio_tmr_start_us for single timer use").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1483,6 +1478,21 @@ void ull_conn_tx_lll_enqueue(struct ll_conn *conn, uint8_t count)
+		curr = memq_peek(curr->next, conn->lll.memq_tx.tail, NULL);
+	}
+
+	return count;
```
Updated.

cvinayak commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -178,6 +182,34 @@ static void write_cmd_cb(struct bt_conn *conn, void *user_data)
+				.interval_min = 0x0028,
+				.interval_max = 0x0028,
+				.latency = 0,
+				.timeout = 30,
```
done

cvinayak committed:
Bluetooth: Controller: Fix assertion establishing connection

Fix an assertion detecting delayed prepare establishing an
ACL connection due to scanner or advertiser not disabled
immediately when scheduling the first connection event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Move declaration close to locality of reference

Move the declaration code to locality of reference and be
consistent in how the code pattern is, related to calling
LLL disable.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Minor update adding comments

Minor update adding comments in the implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix assertion due to cleared pending radio IRQ

Fix assertion on scan disable observed in BabbleSIM audio
tests due to clearing of pending radio IRQ under race
conditions between scanning being stopped by application
and scanning stopped on central connection setup.

irq_enable/disable is not needed either as radio_isr_set
is always called from Radio or LLL execution context which
are co-operative with each other at same execution priority.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley reviewed (changes requested):
I don't like the changes to the GATT samples. I've expressed my opinion on this matter several times, so that shouldn't be a surprise :D 

Now you've added some `CONFIG_TEST` parts to it as well, which is really making the samples look, feel and act more like tests, rather than samples IMO. Consider if you still feel that the GATT sample (which is now being treated as a PHY sample) still follows https://docs.zephyrproject.org/latest/samples/sample_definition_and_criteria.html

Thalley commented on samples/bluetooth/central_gatt_write/src/central_gatt_write.c:
```diff
@@ -123,6 +125,10 @@ uint32_t central_gatt_write(uint32_t count)
 			bt_conn_unref(conn);
 
 			if (count) {
+				if ((count % 1000U) == 0U) {
+					printk("GATT Write countdown %u\n", count);
+				}
```
But what if it's 999?

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -12,11 +12,30 @@
+static uint8_t param_update_idx;
+
+#define CONN_TIMEOUT(_timeout) \
+	BT_GAP_US_TO_CONN_TIMEOUT(DIV_ROUND_UP(MAX(100000U, (_timeout)), \
```
Don't we have a `#define` for that?

Thalley replied:
> No, hci_core.c checks with magic number `10`. Refer to `bt_le_conn_params_valid()`

We should add that then :)

cvinayak replied:
done.

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -12,11 +12,30 @@
+
+#define CONN_TIMEOUT(_timeout) \
+	BT_GAP_US_TO_CONN_TIMEOUT(DIV_ROUND_UP(MAX(100000U, (_timeout)), \
+					       10U * USEC_PER_MSEC) * 10U * USEC_PER_MSEC)
```
How about 
```
define CONN_TIMEOUT(_timeout) \
	MIN(BT_GAP_US_TO_CONN_TIMEOUT(100U * USEC_PER_SEC),  BT_GAP_US_TO_CONN_TIMEOUT(_timeout))
```

That should ensure that the CONN_TIMEOUT value (in units of 10ms) is `_timeout` us (or maximum 100 seconds)

Thalley replied:
> This does not ensure the value is round-up to 10 ms (it will be floored by `BT_GAP_US_TO_CONN_TIMEOUT`

If that's an issue, doesn't that imply that `BT_GAP_US_TO_CONN_TIMEOUT` isn't working optimally? Should `BT_GAP_US_TO_CONN_TIMEOUT` round up, or round to nearest, instead of rounding down?

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -178,6 +205,113 @@ static void write_cmd_cb(struct bt_conn *conn, void *user_data)
+				.latency = 0,
+				.timeout = CONN_TIMEOUT(720000U),
+			}, {
+				.interval_min = BT_GAP_US_TO_CONN_INTERVAL(7500U),
+				.interval_max = BT_GAP_US_TO_CONN_INTERVAL(7500U),
+				.latency = 0,
+				.timeout = CONN_TIMEOUT(720000U),
+			}, {
+				.interval_min = BT_GAP_US_TO_CONN_INTERVAL(50000U),
+				.interval_max = BT_GAP_US_TO_CONN_INTERVAL(50000U),
+				.latency = 0,
+				.timeout = CONN_TIMEOUT(50000U * 6U),
```
> Do we need to fix `bt_le_conn_params_valid()` in the Host?

> The Supervision_Timeout in milliseconds shall be larger than (1 + Max_Latency) × Connection_Interval_Max × 2, where Connection_Interval_Max is given in milliseconds.

I think it is actually correct
```c
	if (param->timeout < 10 || param->timeout > 3200 ||
	    ((param->timeout * 4U) <=
	     ((1U + param->latency) * param->interval_max))) {
		return false;
	}
```

I think it doing some of the conversions between 10ms units and 1.25ms unit implicitly, and thus we have the `* 4`. Probably would be better to have this check more similar to the core spec formula 

> hmmm... I thought the supervision timeout has to be larger than 6 times the connection interval... looks like now it is 

I think the `* 6` is just a good idea to have for a more robust link, right?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -254,13 +280,10 @@ void lll_conn_abort_cb(struct lll_prepare_param *prepare_param, void *param)
-		    lll->periph.window_widening_max_us) {
-			lll->periph.window_widening_prepare_us =
-				lll->periph.window_widening_max_us;
+		lll->periph.window_widening_prepare_us += lll->periph.window_widening_periodic_us *
+							  (prepare_param->lazy + 1);
```
> Typically very large connection intervals with large peripheral latencies or many overlaps do not occur.

Using the word "typically" indicates that it can happen. If that is the case, we need to handle that

jhedberg commented on samples/bluetooth/peripheral_gatt_write/src/peripheral_gatt_write.c:
```diff
@@ -63,6 +79,17 @@ uint32_t peripheral_gatt_write(uint32_t count)
+		printk("Start continuous passive scanning...");
+		err = bt_le_scan_start(BT_LE_SCAN_PASSIVE_ALLOW_DUPILCATES,
+				       device_found);
+		if (err) {
```
This is something we're inconsistent with, probably because we have several Bluetooth maintainers and they have different preferences. I know from a MISRA-C perspective "essentially boolean types" should be used in if-statements, but from a C standard perspective this is actually perfectly fine code. The reason I like it (I'd both produce code like this and also not request someone to change to something else) is that it reads like natural language "if there was an error, then...".

jhedberg replied:
Not sure it's worth it, but a macro like `IS_ERR()` (that expands to a comparison) would still retain natural language style: `if (IS_ERR(err)) ...`

jhedberg replied:
Side note: `if (err < 0)` is actually a bit safer than `if (err != 0)` since we have some APIs where a positive value indicates success (e.g. number of bytes read/written, etc)

jhedberg replied:
>  however, have APIs that return both positive and negative errors 

In a Zephyr context, I'd hope we don't have any of those, especially if the return type is `int` and not some custom typedef, etc. Generally, I think it'd make sense to have a project-wide guideline on the semantics of `int` style error returns, and then we could have helper macros like (`IS_ERR()`) which are documented to be applicable for any errors that adhere to those guidelines.

cvinayak replied:
In the case of `bt_le_scan_start`

```
 * @return Zero on success or error code otherwise, positive in case of
 *         protocol error or negative (POSIX) in case of stack internal error.
```

not sure how much of the case of positive value is true for this API, but particular to this line under review `if (err != 0) {` is correct ... else will be need `IS_ANY_ERR()` / `IS_PROTOCOL_ERR` / `IS_INTERNAL_ERR` etc.!?

jhedberg replied:
I think the core problem here is overloading `int` return value meanings. IMO if we start having anything more complex than "0 for success, negative POSIX for errors" then you shouldn't use `int`. Perhaps `int` is not ideal either, rather there should be some more specific `posix_err_t`. Anyway, we're getting out of scope for this PR. Let's just go with `err != 0` and deal with the rest elsewhere.

Thalley commented on samples/bluetooth/central_gatt_write/src/gatt_write_common.c:
```diff
@@ -178,6 +221,112 @@ static void write_cmd_cb(struct bt_conn *conn, void *user_data)
+				/* No more conn updates, stay at the last index */
+				param_update_iteration = 1U;
+
+				if (IS_ENABLED(CONFIG_BT_OBSERVER) &&
```
Ah, missed that. Somewhat odd to have this in a sample directory called `central_gatt_write`, but I guess it is what it is :D

Thalley reviewed (approved):
Looks OK now, but I will (again) mention that I don't think samples should be used like the `central_gatt_write` and `peripheral_gatt_write` samples are being used now. The samples have become quite complex, and the central sample uses peripheral functions, and vice versa.
